dbt_semantic_interfaces/protocol_example.py

```diff
@@ -0,0 +1,107 @@
+import logging
+from abc import abstractmethod
+from dataclasses import dataclass
+from typing import Protocol, Sequence
+
+from dbt_semantic_interfaces.implementations.base import FrozenBaseModel
+
+logger = logging.getLogger(__name__)
+
+
+# Protocols
+class SemanticModel(Protocol):  # noqa: D
+    @property
+    @abstractmethod
+    def name(self) -> str:  # noqa: D
+        pass
+
+
+class SemanticManifest(Protocol):  # noqa: D
+    @property
+    @abstractmethod
+    def semantic_models(self) -> Sequence[SemanticModel]:  # noqa: D
+        raise NotImplementedError
+
+
+# Explicit Pydantic BaseModel implementations - throws error:
+# TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of
+# all its bases
+
+# class ExplicitPydanticSemanticModel(FrozenBaseModel, SemanticModel):  # noqa: D
+#     name: str
+
+
+# class PydanticSemanticManifest(SemanticManifest, FrozenBaseModel):  # noqa: D
+#     semantic_models: Sequence[PydanticSemanticModel]
+#
+#
+
+
+# Implicit Pydantic implementation
+class ImplicitPydanticSemanticModel(FrozenBaseModel):  # noqa: D
+    name: str
+
+
+class ImplicitPydanticSemanticManifest(FrozenBaseModel):  # noqa: D
+    semantic_models: Sequence[ImplicitPydanticSemanticModel]
```

Comment:
Interesting. I think this works because the protocol only defines the property accessor, not the property setter, so this can be effectively added as a covariant protocol type. That seems like exactly what we need. Assuming setting the properties with ImplicitPydanticSemanticModel subclasses works as expected I think we'll be good to go here.